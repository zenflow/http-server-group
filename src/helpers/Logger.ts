import { PassThrough } from 'stream'

export class Logger extends PassThrough {
  log(string: string) {
    string.split('\n').forEach(line => {
      this.write(`${line}\n`, 'utf8')
    })
  }
}
