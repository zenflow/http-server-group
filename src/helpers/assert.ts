export function assert(value: any, message: string) {
  if (!value) {
    throw new Error(`http-server-group: ${message}`)
  }
}
