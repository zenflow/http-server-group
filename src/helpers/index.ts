export * from './assert'
export * from './createProcess'
export * from './killProcessTree'
export * from './Logger'
export * from './mapStream'
export * from './rightPad'
