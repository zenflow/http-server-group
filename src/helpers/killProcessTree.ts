import treeKill from 'tree-kill'

export function killProcessTree(pid: number): Promise<void> {
  return new Promise((resolve, reject) =>
    treeKill(pid, 'SIGINT', error => (error ? reject(error) : resolve()))
  )
}
