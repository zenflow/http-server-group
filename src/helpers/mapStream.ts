import mapStreamAsync from 'map-stream'
import { Duplex } from 'stream'

export function mapStream(mapper: (arg0: string) => string): Duplex {
  return mapStreamAsync((string: string, cb: Function) =>
    cb(null, mapper(string))
  )
}
