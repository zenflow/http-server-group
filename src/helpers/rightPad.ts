export function rightPad(string: string, length: number): string {
  let result = string
  while (result.length < length) {
    result += ' '
  }
  return result
}
