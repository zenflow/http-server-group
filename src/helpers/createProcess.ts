import { spawn } from 'child_process'
import { Readable } from 'stream'
import { once } from 'events'
import mergeStream from 'merge-stream'
import splitStream from 'split'
import { mapStream } from './mapStream'
import { killProcessTree } from './killProcessTree'

export function createProcess(command: string | Array<string>, env: object) {
  const proc = Array.isArray(command)
    ? spawn(command[0], command.slice(1), {
        env: { ...process.env, ...env },
      })
    : spawn(command, {
        shell: true,
        env: { ...process.env, ...env },
      })
  const outputStream = (mergeStream(
    proc.stdout
      .setEncoding('utf8')
      .pipe(splitStream())
      .pipe(mapStream(line => `[out] ${line}\n`)),
    proc.stderr
      .setEncoding('utf8')
      .pipe(splitStream())
      .pipe(mapStream(line => `[ERR] ${line}\n`))
  ) as unknown) as Readable
  let didExit = false
  const exited = once(outputStream, 'end').then(() => {
    didExit = true
  })
  let killPromise: Promise<void> | null = null
  function kill(): Promise<void> {
    if (!killPromise) {
      killPromise = Promise.resolve().then(async () => {
        if (!didExit) {
          await killProcessTree(proc.pid)
        }
        await exited
      })
    }
    return killPromise
  }
  return { outputStream, exited, kill }
}
