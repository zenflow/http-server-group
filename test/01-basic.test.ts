// @ts-ignore
import { fetchText } from './helpers/fetchText'
import {
  getReadyServerGroupProcess,
  ServerGroupProcess,
  // @ts-ignore
} from './helpers/serverGroupProcess'
// @ts-ignore
import { Config } from '..'

const basicConfig: Config = {
  servers: [
    {
      label: 'a',
      env: { KEY: 'a' },
      command: ['node', 'test/fixtures/server-node-basic.js'],
      port: 3001,
      paths: ['/a'],
    },
    {
      label: 'b',
      env: { KEY: 'b' },
      command: `node test/fixtures/server-node-basic.js`,
      port: 3002,
      paths: ['/b'],
    },
  ],
}

describe('basic', () => {
  jest.setTimeout(30 * 1000)
  let proc: ServerGroupProcess | null = null
  afterEach(async () => {
    if (proc) {
      await proc.kill()
    }
  })
  it('works', async () => {
    proc = await getReadyServerGroupProcess(3000, basicConfig)
    const [aText, bText] = await Promise.all([
      fetchText('http://localhost:3000/a'),
      fetchText('http://localhost:3000/b'),
    ])
    expect(aText).toEqual('a')
    expect(bText).toEqual('b')
  })
  it('has consistent output', async () => {
    proc = await getReadyServerGroupProcess(3000, basicConfig)

    const initialOutput = proc.output.splice(0)
    expect(initialOutput[0]).toBe(`Starting server 'a'...`)
    expect(initialOutput[1]).toBe(`Starting server 'b'...`)
    const aOutStartedLine = initialOutput.indexOf(`a      | [out] Started`)
    expect(aOutStartedLine).toBeGreaterThan(1)
    const bOutStartedLine = initialOutput.indexOf(`b      | [out] Started`)
    expect(bOutStartedLine).toBeGreaterThan(1)
    const startedALine = initialOutput.indexOf(
      `Started server 'a' @ http://localhost:3001`
    )
    expect(startedALine).toBeGreaterThan(aOutStartedLine)
    const startedBLine = initialOutput.indexOf(
      `Started server 'b' @ http://localhost:3002`
    )
    expect(startedBLine).toBeGreaterThan(bOutStartedLine)
    expect(Math.max(startedALine, startedBLine)).toBe(5)
    expect(initialOutput[6]).toBe(`Starting server '$proxy'...`)
    expect(initialOutput[7]).toBe(
      `Started server '$proxy' @ http://localhost:3000`
    )
    expect(initialOutput[8]).toBe('Ready')
    expect(initialOutput[9]).toBeUndefined()

    await proc.kill()

    const finalOutput = proc.output.splice(0)
    expect(finalOutput[0]).toBe('')
    expect(finalOutput[1]).toBe('')
  })
})
