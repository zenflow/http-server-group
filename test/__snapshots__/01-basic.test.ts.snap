// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic has consistent output 1`] = `
Object {
  "": Array [
    "Starting 'a'...",
    "Starting 'b'...",
    "Started 'a' @ http://localhost:3001",
    "Started 'b' @ http://localhost:3002",
    "Starting '$proxy'...",
    "Started '$proxy' @ http://localhost:3000",
    "Ready",
  ],
  "$proxy": Array [],
  "a": Array [
    "a      | [out] Starting...",
    "a      | [out] Started",
  ],
  "b": Array [
    "b      | [out] Starting...",
    "b      | [out] Started",
  ],
}
`;

exports[`basic has consistent output 2`] = `
Array [
  "",
  "",
]
`;
